[FLINK-31644][network] Implement the disk tier producer for the tiered storage

TanYuxin-tyx opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*Introduce the partition file writer and Implement the disk tier producer for the tiered storage*


## Brief change log

  - *【commit 1】 Introduce the partition file writer interface for the tiered storage*
  - *【commit 2】Introduce the interface and implementation for the partition file index*
  - *【commit 3】Implement ProducerMergePartitionFileWriter for the tiered storage*
  - *【commit 4】Implement the disk tier producer for the tiered storage*


## Verifying this change

This change added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yno)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,109 @@
+     *
+     * @param spilledBuffers the buffers to be spilled
+     */
+    void addRegionForBuffers(List<SpilledBuffer> spilledBuffers);
```
The contract here is ambiguous. What does this method do? What are the relation ship between regions and the provided buffers?

TanYuxin-tyx replied:
1. Rename the method to `generateRegionsBasedOnBuffers` and added detailed doc about the method.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,109 @@
+     * @param nettyServiceWriterId the netty connection which reads the buffers from the partition
+     *     file, then writes the buffers to the netty.
+     */
+    Optional<Region> getNextRegion(int subpartitionId, NettyConnectionId nettyServiceWriterId);
```
Why is there a netty connection id?

xintongsong replied:
Who's calling this? Why would the index remembers which region it has read previously?

TanYuxin-tyx replied:
We have refactored the method and it does not need `NettyConnectionId` anymore.
When reading the data from the disk, the subpartition reader will call the method.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,109 @@
+ * the subpartition, the offset in file it begin with. The {@link Region} represents a series of
+ * physically continuous buffers in the file, which are from the same subpartition.
+ */
+public interface PartitionFileIndex {
```
Which component is this interface designed to be used by? What different implementations can it possibly have?

TanYuxin-tyx replied:
Removed the API of PartitionFileIndex.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/PartitionFileIndexImpl.java:
```diff
@@ -0,0 +1,159 @@
+            if (currentBuffer.getSubpartitionId() != firstBufferInRegion.getSubpartitionId()
+                    || currentBuffer.getBufferIndex() != lastBufferInRegion.getBufferIndex() + 1) {
+                // the current buffer belongs to a new region, close the previous region
+                addInternalRegionToMap(
+                        firstBufferInRegion, lastBufferInRegion, subpartitionRegionMap);
```
It seems we require `spilledBuffers` to be continuous, which is never mentioned in the contract.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/ProducerMergePartitionFileWriter.java:
```diff
@@ -0,0 +1,182 @@
+
+    @Override
+    public CompletableFuture<Void> write(List<SubpartitionNettyPayload> toWriteBuffers) {
+        List<NettyPayload> buffersToSpill =
+                toWriteBuffers.stream()
+                        .map(SubpartitionNettyPayload::getSegmentNettyPayloads)
+                        .flatMap(
+                                (Function<List<SegmentNettyPayload>, Stream<SegmentNettyPayload>>)
+                                        Collection::stream)
+                        .map(SegmentNettyPayload::getNettyPayloads)
+                        .flatMap(
+                                (Function<List<NettyPayload>, Stream<NettyPayload>>)
+                                        Collection::stream)
+                        .collect(Collectors.toList());
```
Then why do we need the complicated `SubpartitionNettyPayload` and `SegmentNettyPayload` at the first place?

TanYuxin-tyx replied:
Because the subpartition id and segment id are used in the hash mode partition file. While the producer merge mode does not need that.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:
```diff
@@ -0,0 +1,157 @@
+    public void appendEndOfSegmentEvent(
+            ByteBuffer record, int subpartitionId, Buffer.DataType dataType) {
+        subpartitionCacheManagers[subpartitionId].appendEndOfSegmentEvent(record, dataType);
+        flushAndReleaseCacheBuffers();
```
Why do we need flush here?

TanYuxin-tyx replied:
Added a detailed doc about the flush.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:
```diff
@@ -0,0 +1,157 @@
+     * @return the finished buffer index
+     */
+    public int getFinishedBufferIndex(int subpartitionId) {
+        return subpartitionCacheManagers[subpartitionId].getFinishedBufferIndex();
```
Which buffer is the finished buffer?

TanYuxin-tyx replied:
I have removed the concept of `finished`. Rename it to `BufferIndex` directly.

Each time a new buffer is added, the field of `BufferIndex` will increase by 1.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:
```diff
@@ -0,0 +1,157 @@
+                    new SubpartitionNettyPayload(
+                            subpartitionId,
+                            Collections.singletonList(
+                                    new SegmentNettyPayload(-1, nettyPayloads, false))));
```
Why is the hardcoded `-1` segment id and `false` for finishing segment?

TanYuxin-tyx replied:
Replaced the `-1` by the right segment id.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/SubpartitionDiskCacheManager.java:
```diff
@@ -0,0 +1,109 @@
+    List<NettyPayload> getAllBuffers() {
+        synchronized (allBuffers) {
+            List<NettyPayload> targetBuffers = new ArrayList<>(allBuffers);
+            allBuffers.clear();
```
Neither the name nor any javadoc has suggested that the buffers will be cleared.

xintongsong reviewed (commented):
It's probably not a good practice to separate the reading & writing parts of the partition file into different commits / PRs. E.g., it's hard for me to say whether the `PartitionFileIndex` is well defined, because I don't know how it is used.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/PartitionFileWriter.java:
```diff
@@ -0,0 +1,41 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.file;
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/file/SegmentNettyPayload.java:
```diff
@@ -0,0 +1,55 @@
+ * The wrapper class {@link SegmentNettyPayload} for a segment, which holds all the {@link
+ * NettyPayload} buffers and the flag that mark whether this segment need to be finished.
+ */
+public class SegmentNettyPayload {
```
1. Added a new class `SubpartitionSpilledBufferContext` and `SegmentSpilledBufferContext` in `PartitionFileWriter`.
2. Removed the `SubpartitionNettyPayload ` and `SegmentNettyPayload `

TanYuxin-tyx commented:
@xintongsong Thanks for reviewing.
I have refactored the commits of the PR. The partition file reader  API is also submitted. PTAL

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileWriter.java:
```diff
@@ -0,0 +1,107 @@
+     * @return the completable future indicating whether the writing file process has finished. If
+     *     the {@link CompletableFuture} is completed, the written process is completed.
+     */
+    CompletableFuture<Void> write(List<SubpartitionSpilledBufferContext> spilledBuffers);
```
Fixed the comments and renamed the argument to `buffersToWrite`

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileWriter.java:
```diff
@@ -0,0 +1,107 @@
+        private final List<SpilledBufferContext> spilledBufferContexts;
+
+        /** Whether it is necessary to finish the segment. */
+        private final boolean needFinishSegment;
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SpilledBufferContext.java:
```diff
@@ -0,0 +1,65 @@
+    private final int subpartitionId;
+
+    /** The id of segment. */
+    private final int segmentId;
```
Removed the class `SpilledBufferContext`

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileReader.java:
```diff
@@ -0,0 +1,53 @@
+            long fileOffset,
+            MemorySegment memorySegment,
+            BufferRecycler recycler)
+            throws IOException;
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileReader.java:
```diff
@@ -0,0 +1,53 @@
+     * @param subpartitionId subpartition id indicates the id of subpartition.
+     * @param fileOffset file offset indicates the current reading offset.
+     * @param memorySegment memory segment indicates a segment of available buffer.
+     * @param recycler recycler indicates the owner of the buffer.
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileReader.java:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @param partitionId partition id indicates the id of partition.
+     * @param subpartitionId subpartition id indicates the id of subpartition.
+     * @param fileOffset file offset indicates the current reading offset.
```
Fixed with a `bufferIndex`.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,210 @@
+/**
+ * The {@link PartitionFileIndex} represents the indexes and the regions of the spilled buffers. The
+ * region indexes are generated when writing the spilled buffers, and these region indexes are used
+ * when reading data from disk.
```
Re-write the doc.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/SubpartitionDiskCacheManager.java:
```diff
@@ -0,0 +1,122 @@
+
+    // Note that this field can be accessed by the task thread or the write IO thread, so the thread
+    // safety should be ensured.
+    private final Deque<NettyPayload> allBuffers = new LinkedList<>();
```
Removed. Replace that with `Tuple2<Buffer, Integer>`, which is buffer and bufferIndex.

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/SpilledBufferContext.java:
```diff
@@ -0,0 +1,65 @@
+ * The buffer context to be flushed, including the {@link Buffer}, the buffer index, the
+ * subpartition id, the segment id, etc.
+ */
+public class SpilledBufferContext {
```
`SpilledBufferContext` is removed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileReader.java:
```diff
@@ -0,0 +1,73 @@
+     * @param segmentId the segment id of the buffer
+     * @param bufferIndex the index of buffer
+     * @return the priority of the {@link PartitionFileReader}.
+     */
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,218 @@
+ * process, the {@link Region}s are generated based on the buffers. During partition file reads, the
+ * {@link Region} is used to retrieve consecutive buffers that belong to a single subpartition.
+ */
+public class PartitionFileIndex {
```
Renamed the filed index and the writer/reader API.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,218 @@
+ * file, these indexes are encapsulated into a {@link Region}. During the partition file write
+ * process, the {@link Region}s are generated based on the buffers. During partition file reads, the
+ * {@link Region} is used to retrieve consecutive buffers that belong to a single subpartition.
+ */
```
Fixed the commentes.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,218 @@
+    // ------------------------------------------------------------------------
+
+    /** Represents a buffer to be flushed. */
+    public static class BufferToFlush {
```
Rename this to `FlushedBuffer` because it represents only one buffer.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,218 @@
+     *
+     * @param bufferToFlushes the buffers to be flushed
+     */
+    void generateRegionsBasedOnBuffers(List<BufferToFlush> bufferToFlushes) {
```
Renamed the method.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,218 @@
+     * they remain physically contiguous when flushing to disk.
+     *
+     * @param bufferToFlushes the buffers to be flushed
+     */
```
Fixed the doc.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,218 @@
+     * @param subpartitionId the specific subpartition id
+     * @param regionId the region id to get from the {@link PartitionFileIndex}
+     */
+    public Optional<Region> getRegion(int subpartitionId, int regionId) {
```
Replaced the `regionId` with `bufferIndex`

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileIndex.java:
```diff
@@ -0,0 +1,218 @@
+    /**
+     * A {@link Region} represents a series of physically continuous buffers in the file, which are
+     * from the same subpartition.
+     */
```
Fixed the docs.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/ProducerMergePartitionFile.java:
```diff
@@ -0,0 +1,37 @@
+ * The partition file in the producer-merge mode. In this mode, the shuffle data is written in the
+ * producer side, the consumer side need to read multiple producers to get its partition data.
+ */
+public class ProducerMergePartitionFile {
```
Renamed the classes of partition file, file writer, file reader, file index.

TanYuxin-tyx commented:
@xintongsong Thanks for reviewing.
After addressing the comments, I checked the PR overall and made some fixes.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:
```diff
@@ -0,0 +1,174 @@
+    private final SubpartitionDiskCacheManager[] subpartitionCacheManagers;
+
+    /** Whether the current flush process has completed. */
+    private CompletableFuture<Void> hasFlushCompleted;
```
```suggestion
    private CompletableFuture<Void> onGoingFlushFuture;
```

TanYuxin-tyx replied:
Renamed and added docs to describe this field.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:
```diff
@@ -0,0 +1,174 @@
+     * Note that the request of flushing buffers may come from the disk check thread or the task
+     * thread, so the method itself should ensure the thread safety.
+     */
+    private synchronized void flushBuffers(boolean needForceFlush) {
```
```suggestion
    private synchronized void flushBuffers(boolean forceFlush) {
```

TanYuxin-tyx replied:
fixed.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java:
```diff
@@ -0,0 +1,174 @@
+     * @param subpartitionId target subpartition of this record.
+     * @param dataType the type of this record. In other words, is it data or event.
+     */
+    void appendEndOfSegmentEvent(ByteBuffer record, int subpartitionId, Buffer.DataType dataType) {
```
`Buffer.DataType` dataType seems can be removed safety.

TanYuxin-tyx replied:
This should not be removed, because the previous comments show that the concept of end-of-segment need not be aware both in `SubpartitionDiskCacheManager` and `DiskProducerAgent`. I also think this makes sense. 

So the `Buffer.DataType` is to avoid composing the end-of-segment to `SubpartitionDiskCacheManager`.

reswqa replied:
What confuses me is: this method called `appendEndOfSegmentEvent`, Isn't this implying that it must be an event? Is it possible that it will be `Data` in the future? 

If we have a method called `emitEvent (Buffer event)` in a class `A`, I don't think this can be considered as `A` exactly aware the specific event.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/SubpartitionDiskCacheManager.java:
```diff
@@ -0,0 +1,130 @@
+        addBuffer(new NetworkBuffer(data, FreeingBufferRecycler.INSTANCE, dataType, data.size()));
+    }
+
+    /** Note that allBuffers can be touched by multiple threads. */
```
Is this method only called by task thread?

TanYuxin-tyx replied:
Fixed.

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/SubpartitionDiskCacheManager.java:
```diff
@@ -0,0 +1,130 @@
+     * Record the segment id that is writing to. Each time when the segment is finished, this filed
+     * is increased by one.
+     */
+    private int segmentIndex;
```
I'm a little suspicious that this field is unsafe.

TanYuxin-tyx replied:
Guarded it by `allBuffers`.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/PartitionFileWriter.java:
```diff
@@ -0,0 +1,112 @@
+            return bufferAndIndexes;
+        }
+
+        public boolean isSegmentFinished() {
```
This method is designed to be used in the hash mode partition file.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/ProducerMergedPartitionFileIndex.java:
```diff
@@ -0,0 +1,246 @@
+            this.numBuffers = numBuffers;
+        }
+
+        long getRegionFileOffset() {
```
This method is designed to be used when reading data.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/file/ProducerMergedPartitionFileIndex.java:
```diff
@@ -0,0 +1,246 @@
+            return regionFileOffset;
+        }
+
+        int getNumBuffers() {
```
This method is designed to be used when reading data.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/common/TieredStorageUtils.java:
```diff
@@ -0,0 +1,52 @@
+/** Utils for reading or writing from tiered storage. */
+public class TieredStorageUtils {
+
+    public static final String DATA_FILE_SUFFIX = ".storage.data";
```
Fixed.

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/SubpartitionDiskCacheManager.java:
```diff
@@ -0,0 +1,130 @@
+     * Record the buffer index in the {@link SubpartitionDiskCacheManager}. Each time a new buffer
+     * is added to the {@code allBuffers}, this field is increased by one.
+     */
+    private int bufferIndex;
```
Fixed and added the doc.

TanYuxin-tyx commented:
@reswqa Thanks for the review. Note that I addressed the comments with a fixup commit only for being convenient to review the change.
After the review, the files(e.g., the test classes) in the fixup commit will be merged into the previous commits according to their production code.

reswqa commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/SubpartitionDiskCacheManagerTest.java:
```diff
@@ -0,0 +1,81 @@
+        SubpartitionDiskCacheManager subpartitionDiskCacheManager =
+                new SubpartitionDiskCacheManager();
+        subpartitionDiskCacheManager.append(BufferBuilderTestUtils.buildSomeBuffer());
+        assertThatNoException().isThrownBy(subpartitionDiskCacheManager::release);
```
We'd better check this specific buffer is recycled.

TanYuxin-tyx replied:
Added the check.

TanYuxin-tyx commented:
@reswqa Thanks for review, I have addressed the comments, PTAL, thanks.
